lib/geckodriver.js

```diff
@@ -0,0 +1,192 @@
+
+  getVersion = _.memoize(async () => {
+    const proc = new SubProcess(this.geckodriverExecutable, ['--version']);
+    let version;
+    await proc.start((stdout) => {
+      log.debug('Checking geckodriver version from stdout:', stdout);
+      version = this.parseVersion(stdout);
+      return true;
+    });
+    return version;
+  });
+
+  async start (caps, emitStartingState = true) {
+    this.capabilities = _.cloneDeep(caps.desiredCapabilities || caps.capabilities);
+
+    const args = [`--host=${this.proxyHost}`, `--port=${this.proxyPort}`, '-vv'];
+
+    const startDetector = (stdout) => {
+      return stdout.includes('Listening');
+    };
+
+    if (emitStartingState) {
+      this.changeState(Geckodriver.STATE_STARTING);
+    }
+
+    let version = await this.getVersion();
+    log.debug(`Geckodriver version ${version}`);
+    if (util.compareVersions(version, '<', MIN_SUPPORTED_VERSION)) {
+      log.errorAndThrow(`Only geckodriver ${MIN_SUPPORTED_VERSION} supports Android testing. Current version: ${version}`);
+    }
+
+    let processIsAlive = false;
+    try {
+      this.proc = new SubProcess(this.geckodriverExecutable, args);
+      processIsAlive = true;
+
+      this.proc.on('output', (stdout, stderr) => {
+        if (this.verbose) {
+          for (let line of (stdout || '').trim().split('\n')) {
+            if (!line.trim().length) continue; // eslint-disable-line curly
+            log.debug(`[STDOUT] ${line}`);
+          }
+          for (let line of (stderr || '').trim().split('\n')) {
+            if (!line.trim().length) continue; // eslint-disable-line curly
+            log.error(`[STDERR] ${line}`);
+          }
+        }
+      });
+
+      this.proc.on('exit', (code, signal) => {
+        log.info(code, signal);
+        processIsAlive = false;
+
+        if (![Geckodriver.STATE_STOPPED, Geckodriver.STATE_STOPPING, Geckodriver.STATE_RESTARTING].includes(this.state)) {
+          let msg = `Geckodriver exited unexpectedly with code ${code}, ` +
+                  `signal ${signal}`;
+          log.error(msg);
+          this.changeState(Geckodriver.STATE_STOPPED);
+        }
+      });
+
+      log.info(`Spawning geckodriver with: ${quote([this.geckodriverExecutable, ...args])}`);
+      // start subproc and wait for startDetector
+      await this.proc.start(startDetector);
+      this.changeState(Geckodriver.STATE_ONLINE);
+      await this.startSession();
+    } catch (e) {
+      this.emit(Geckodriver.EVENT_ERROR, e);
+      if (processIsAlive) {
+        await this.stop();
+      }
+    }
+  }
+
+  sessionId () {
+    if (this.state !== Geckodriver.STATE_ONLINE) {
+      return null;
+    }
+    return this.jwproxy.sessionId;
+  }
+
+  async stop () {
+    try {
+      await this.proc.stop();
+      this.changeState(Geckodriver.STATE_STOPPED);
+    } catch (e) {
+      this.emit(Geckodriver.EVENT_ERROR, e);
+    }
+  }
+
+  async deleteSession () {
+    await this.jwproxy.command('', 'DELETE', {});
```

Comment:
shouldn't we also stop the driver if it is running?

Reply:
That would actually fix the issue im seeing since `DELETE` is not working. So any new session would start a new child process for the driver?

Reply:
The new version of Firefox fixed the `DELETE` issue. Gonna add that we stop the driver on deleteSession.